File: 05_Paths_in_Labyrinth.cpp
#include <iostream>
#include <vector>

void dfs(std::vector<std::vector<char>>& labyrinth, std::vector<std::vector<bool>>& visited, int x, int y, std::string& path) {
    int rows = labyrinth.size();
    int cols = labyrinth[0].size();

    if (labyrinth[x][y] == 'e') {
        std::cout << path << std::endl;
        return;
    }

    visited[x][y] = true;

    std::vector<std::pair<int, int>> directions = {{0, 1}, {1, 0}, {0, -1}, {-1, 0}};
    std::vector<char> direction_symbols = {'R', 'D', 'L', 'U'};

    for (int i = 0; i < directions.size(); i++) {
        int new_x = x + directions[i].first;
        int new_y = y + directions[i].second;

        if (new_x >= 0 && new_x < rows && new_y >= 0 && new_y < cols && !visited[new_x][new_y] && labyrinth[new_x][new_y] != '*') {
            path.push_back(direction_symbols[i]);
            dfs(labyrinth, visited, new_x, new_y, path);
            path.pop_back();
        }
    }

    visited[x][y] = false;
}

void find_all_paths(std::vector<std::vector<char>>& labyrinth) {
    int rows = labyrinth.size();
    int cols = labyrinth[0].size();
    std::vector<std::vector<bool>> visited(rows, std::vector<bool>(cols, false));
    std::string path;

    dfs(labyrinth, visited, 0, 0, path);
}

int main() {
    int rows, cols;

    std::cout << "Enter the number of rows: ";
    std::cin >> rows;
    std::cout << "Enter the number of columns: ";
    std::cin >> cols;

    std::cout << "Enter the labyrinth:" << std::endl;
    std::vector<std::vector<char>> labyrinth(rows, std::vector<char>(cols));

    for (int i = 0; i < rows; i++) {
        std::string row;
        std::cin >> row;
        for (int j = 0; j < cols; j++) {
            labyrinth[i][j] = row[j];
        }
    }

    std::cout << "Paths to the exit:" << std::endl;
    find_all_paths(labyrinth);

    return 0;
}

File: 05_Paths_in_Labyrinth.py
def find_all_paths(labyrinth):
    rows, cols = len(labyrinth), len(labyrinth[0])
    visited = [[False]*cols for _ in range(rows)]
    directions = [(0, 1, 'R'), (1, 0, 'D'), (0, -1, 'L'), (-1, 0, 'U')]

    def is_valid(x, y):
        if x < 0 or y < 0 or x >= rows or y >= cols:
            return False
        if visited[x][y] or labyrinth[x][y] == '*':
            return False
        return True

    def dfs(x, y, path):
        if labyrinth[x][y] == 'e':
            print(path)
            return
        visited[x][y] = True
        for dx, dy, direction in directions:
            new_x, new_y = x + dx, y + dy
            if is_valid(new_x, new_y):
                dfs(new_x, new_y, path + direction)
        visited[x][y] = False

    dfs(0, 0, '')

# Въвеждане на брой редове и колони от конзолата
rows = int(input("Въведете брой редове: "))
cols = int(input("Въведете брой колони: "))

# Въвеждане на лабиринта от конзолата
labyrinth = []
for _ in range(rows):
    row = list(input("Въведете ред от лабиринта: "))
    labyrinth.append(row)

# Извикване на функцията за намиране на всички пътища
find_all_paths(labyrinth)
